docs/root/faq/rolling_upgrades.rst

```diff
@@ -0,0 +1,124 @@
+
+* *CANCELLED* - Return this code if there is an error that can be retried in the service.
+* *RESOURCE_EXHAUSTED* - Return this code if some of the resources that service depends on are exhausted in that instance so that retrying 
+  to another instance would help. Please note that for shared resource exhaustion, returning this will not help. Instead :ref:`rate limiting<arch_overview_rate_limit>`
+  should be used to handle such cases.
+
+The Http Status codes *502 (Bad Gateway)*, *503 (Service Unavailable)* and *504 (Gateway Timeout)* are all mapped to gRPC status code *UNAVAILABLE*. 
+This can also be considered safe for automatic retry.
+
+An important consideration should be given to *idempotency* of a request while configuring retries.
+
+Envoy also supports various extensions to the standard retry policy supported out of the box. 
+The :ref:`retry plugins<arch_overview_http_retry_plugins>` allow you to customize the Envoy 
+retry implementation to your application.
+
+Outlier Detection
+-----------------
+
+:ref:`Outlier detection<arch_overview_outlier_detection>` is a way of dynamically determining outlier 
+hosts in the upstream cluster. If certain hosts behave differently from others like returning high 5xx 
+or success rate differs  from other hosts, etc. By detecting such hosts and ejecting them for a 
+temporary period of time from the healthy load balancing set, Envoy can increase the success rate of a request. 
+Envoy supports configuring outlier detection based on continuous 5xx, continuous gateway failures and success rate.
+
+Envoy also allows you to configure the ejection period.
+
+**Configuration**
+
+The following settings help to optimize some combination of:
+
+* Maximum request success for common scenarios (i.e. rolling upgrade) 
+* Speed 
+* Avoid cascading failures
+
+
+*Circuit Breaker*
+
+.. code-block:: json
+
+  {
+     "thresholds": [
+       {
+         "max_retries": 10,
+       }
+    ]
+  }
+
+For the purpose of this specific use case, the retry budget for upstream cluster should be configured to
+enable and control concurrent retries. If the value configured is too low, some requests will not be retried 
+and will result in :ref:`upstream_rq_retry_overflow <config_cluster_manager_cluster_stats>`
+and if the value is configured is too high, the service can be overwhelmed with retry requests. 
+
+
+*Outlier Detection*
+
+.. code-block:: json
+
+  {
+     "consecutive_5xx": 5,
+     "base_ejection_time": "30s",
+     "max_ejection_percent": 50,
+     "consecutive_gateway_failure": 5,
+  }
+
+This setting enables outlier detection if there are 5 consecutive *5xx* or *gateway failures* and limit the number of hosts that are ejected to 50% of the upstream cluster size. 
+This will allow us to keep the available hosts in the load balancer to be optimum for retries while ejecting truly bad hosts. Please note that once a host a ejected, it would be brought back 
```

Comment:
I'm not sure what "optimium for retries" means. I think the config places a safe limit on the number of hosts removed.

Replace "it would be brought back in to" with "it will be returned to"